### PR TITLE
Filling the polyomino with vertical and horizontal lines

### PR DESCRIPTION
**Pull request by acanalpay**
Filling the polyomino's inside to prevent overlapping when performing layout operation.
To fill the polyomino, code simply finds the left-most and right-most occupied cells for x-axes, also finds the upper-most and lower-most cells on the y-axes, and fills the cells between them.

### CODE REVIEW

**Comment by hasanbalci** [comment]
@acanalpay Thanks, it seems like a clean solution. I only noticed one case where polyominos aren't filled correctly: 

layout-utilities-bug

If a vertical/horizontal line enters/exits multiple times to/from the polyomino during sweep operation, it fills all grids the line passes. For example, in this case, region 2 is also filled, so small component cannot enter there.

**Comment by acanalpay** [comment]
> @acanalpay Thanks, it seems like a clean solution. I only noticed one case where polyominos aren't filled correctly:
> 
> layout-utilities-bug
> 
> If a vertical/horizontal line enters/exits multiple times to/from the polyomino during sweep operation, it fills all grids the line passes. For example, in this case, region 2 is also filled, so small component cannot enter there.

Thanks for the tests. Currently, I only look for the left-most and right-most and fill in between. I'll modify the solution for such cases to avoid filling region 2

**Comment by ugurdogrusoz** [comment]
So, is the idea is to fill in between *pairs* of occupied cells from left to right and fill only in between them? Between 1 & 2, 3 & 4, etc.?

**Comment by acanalpay** [comment]
> So, is the idea is to fill in between _pairs_ of occupied cells from left to right and fill only in between them? Between 1 & 2, 3 & 4, etc.?

Algorithm finds left-most right-most and top-most bottom-most and fills in between for every line. In this case left-most cell is located in n5-n6 edge and right-most cell is in n7-n10 edge. So, regardless of n6-n11 and n11-n10 edges it will fill everything. Similarly, it fills from n5-n7 edge to n6-n11 and n11-n10 edges.

With this solution I cannot avoid filling the region 2. I implemented a possible solution but it won't work for some cases. The approach of the solution for this case is, while filling for left to right, if the polyomino.grid is already occupied, then it is on the n6-n11 edge. So, algorithm stops filling until the next occupied polyomino.grid cell (which is on the edge n11-n10). Then I start filling again etc.. This approach seems to work, however, this solution gives wrong filling when multiple edges located inside the graph (like the one in the image).  

image

I'll find a more suitable solution than filling with horizontal and vertical lines to avoid filling region 2.

**Comment by acanalpay** [comment]
Bugs are fixed and the documentation with the running time tests can be found [here](https://docs.google.com/document/d/1Ka-0TMtVf6x2n16185SZ-C3ZqAiKPPbndO2ktX2oiDA/edit?usp=sharing)

**Comment by acanalpay** [comment]
The solution didn't fully fix the adjusted fullness problem.  While calculating adjusted fullness, the imaginary edges are counted as normal edges, resulting in a slightly higher ratio.  

Imaginary edges: Edges that the algorithm draw to find the smallest face covering polyomino

Normal edges: The edges that are already presented in the polyomino (the ones that we see)

**Comment by ugurdogrusoz** [comment]
@hasanbalci Please review and merge to unstable branch if things look ok. Tests looked OK to me.

File: src/core/layout-utilities.js [hunk 1]
```diff
@@ -350,8 +350,10 @@ var layoutUtilities = function (cy, options) {
 
   /**
    * @param { any[] } components 
+   * @param {boolean} spacewise include the empty cells inside the components to calculate how well the algorithm perform
+   * spacewise is false by default so the algorithm looks how well the free cells (outside the polyomino) is used
```
Comment (hasanbalci) [inline src/core/layout-utilities.js]: I think you mean "... spacewise is true by default..." based on the default parameter below.

**Review by hasanbalci — CHANGES_REQUESTED** [review]
@acanalpay Thanks for the great effort! Finally, could you please update the README.md since you added a new parameter `spacewise` to the `packComponents` function in the API?

**Review by acanalpay — COMMENTED** [review]
readme file is modified

**Comment by ugurdogrusoz** [comment]
How about renaming spacewise  to considerAsPolygons?

**Pull request merged**